pkg/controller/unifiedpushserver/unifiedpushserver.go

```diff
@@ -75,26 +100,63 @@ func newUnifiedPushServerDeployment(cr *pushv1alpha1.UnifiedPushServer) (*appsv1
 		return nil, errors.Wrap(err, "error generating cookie secret")
 	}
 
-	return &appsv1.Deployment{
+	return &openshiftappsv1.DeploymentConfig{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      cr.Name,
 			Namespace: cr.Namespace,
 			Labels:    labels,
 		},
-		Spec: appsv1.DeploymentSpec{
-			Selector: &metav1.LabelSelector{
-				MatchLabels: labels,
+		Spec: openshiftappsv1.DeploymentConfigSpec{
+			Replicas: 1,
+			Selector: labels,
+			Triggers: openshiftappsv1.DeploymentTriggerPolicies{
+				openshiftappsv1.DeploymentTriggerPolicy{
+					Type: openshiftappsv1.DeploymentTriggerOnConfigChange,
+				},
+				openshiftappsv1.DeploymentTriggerPolicy{
+					Type: openshiftappsv1.DeploymentTriggerOnImageChange,
+					ImageChangeParams: &openshiftappsv1.DeploymentTriggerImageChangeParams{
+						Automatic:      true,
+						ContainerNames: []string{cfg.UPSContainerName},
+						From: corev1.ObjectReference{
+							Kind: "ImageStreamTag",
+							Name: cfg.UPSImageStreamName + ":" + cfg.UPSImageStreamTag,
+						},
+					},
+				},
+				openshiftappsv1.DeploymentTriggerPolicy{
+					Type: openshiftappsv1.DeploymentTriggerOnImageChange,
+					ImageChangeParams: &openshiftappsv1.DeploymentTriggerImageChangeParams{
+						Automatic:      true,
+						ContainerNames: []string{cfg.PostgresContainerName},
+						From: corev1.ObjectReference{
+							Kind: "ImageStreamTag",
+							Name: cfg.PostgresImageStreamName + ":" + cfg.PostgresImageStreamTag,
```

Comment:
Is it correct to trigger this UPS DeploymentConfig on an updated PostgreSQL ImageStreamTag update, or should this one instead be in the PostgreSQL DeploymentConfig (I notice there aren't any triggers in there)?

Reply:
> Is it correct to trigger this UPS DeploymentConfig on an updated PostgreSQL ImageStreamTag update

That's a preference IMO. Happy to delete that trigger for UPS.

> or should this one instead be in the PostgreSQL DeploymentConfig 

I will check it. Perhaps I forgot it.

Reply:
> Is it correct to trigger this UPS DeploymentConfig on an updated PostgreSQL ImageStreamTag update

Sent an email to AeroGear ML about this: https://groups.google.com/forum/#!topic/aerogear/D7vncataJac

Reply:
> Is it correct to trigger this UPS DeploymentConfig on an updated PostgreSQL ImageStreamTag update

When I don't have this here, the init container fails to resolve the image from the imageStream. I guess this would be possible but I couldn't figure out how. Let's pair on this

Reply:
That's interesting! :thinking: 

Does it still work if you set `Automatic: false`? If the trigger is required (which it sounds like it is) then it sounds like in our case setting automatic to be false might help. Unfortunately the API doesn't seem to specify what the default value of that flag is: https://github.com/openshift/origin/blob/8da9ab8f033427d7d463308fc7035e95fd20a7b4/staging/src/github.com/openshift/openshift-apiserver/pkg/apps/apis/apps/types.go#L277

Reply:
Tried `Automatic=false` already. Doesn't work. It simply doesn't do the deployment of the init proxy in that case (ie. no pods created with a message like "waiting for something to happen".)

Reply:
Hmm, okay, in that case I think we have two options, and either are fine:

- allow it to trigger the UPS deployment when there's a new Postgresql image, and that's fine because the postgresql will be redeployed anyway, so UPS won't be available while that's happening.
- Don't use an ImageStream for this init container: it seems more correct to not trigger for this than to trigger, since in a successfully running pod this image isn't running. If we leave it at a floating tag, with an ImagePullPolicy set to Always, I think we get the behaviour that we want (no unexpected redeploy, use the newest image the next time it's needed).

What do you think?

Reply:
IMO, option 1 is better. Otherwise there will be more confusion around the Postgres image we use. I personally wouldn't like it when there is an ImageStream we use for Postgres as well as a regular Image.